flow/display_list.cc

```diff
@@ -0,0 +1,1248 @@
+    static const auto kType = DisplayListOpType::Draw##name;                 \
+                                                                             \
+    Draw##name##Op(DRAW_ATLAS_##cull##_ARGS) : DRAW_ATLAS_##cull##_INIT {}   \
+                                                                             \
+    DRAW_ATLAS_##cull##_FIELDS;                                              \
+                                                                             \
+    void dispatch(Dispatcher& dispatcher) const {                            \
+      const SkRSXform* xform = reinterpret_cast<const SkRSXform*>(this + 1); \
+      const SkRect* tex = reinterpret_cast<const SkRect*>(xform + count);    \
+      const SkColor* colors = DRAW_ATLAS_##colors##_ARRAY(tex, count);       \
+      dispatcher.drawAtlas(atlas, xform, tex, colors, count, mode, sampling, \
+                           DRAW_ATLAS_##cull##_P_ARG);                       \
+    }                                                                        \
+  };
+DEFINE_DRAW_ATLAS_OP(Atlas, NO_COLORS, NO_CULLING)
+DEFINE_DRAW_ATLAS_OP(AtlasColored, HAS_COLORS, NO_CULLING)
+DEFINE_DRAW_ATLAS_OP(AtlasCulled, NO_COLORS, HAS_CULLING)
+DEFINE_DRAW_ATLAS_OP(AtlasColoredCulled, HAS_COLORS, HAS_CULLING)
+#undef DEFINE_DRAW_ATLAS_OP
+#undef DRAW_ATLAS_NO_COLORS_ARRAY
+#undef DRAW_ATLAS_HAS_COLORS_ARRAY
+#undef DRAW_ATLAS_NO_CULLING_ARGS
+#undef DRAW_ATLAS_NO_CULLING_INIT
+#undef DRAW_ATLAS_NO_CULLING_FIELDS
+#undef DRAW_ATLAS_NO_CULLING_P_ARG
+#undef DRAW_ATLAS_HAS_CULLING_ARGS
+#undef DRAW_ATLAS_HAS_CULLING_INIT
+#undef DRAW_ATLAS_HAS_CULLING_FIELDS
+#undef DRAW_ATLAS_HAS_CULLING_P_ARG
+
+// 4 byte header + ptr aligned payload uses 12 bytes rounde up to 16
+// (4 bytes unused)
+struct DrawSkPictureOp final : DLOp {
+  static const auto kType = DisplayListOpType::DrawSkPicture;
+
+  DrawSkPictureOp(sk_sp<SkPicture> picture) : picture(std::move(picture)) {}
+
+  sk_sp<SkPicture> picture;
+
+  void dispatch(Dispatcher& dispatcher) const {
+    dispatcher.drawPicture(picture);
+  }
+};
+
+// 4 byte header + ptr aligned payload uses 12 bytes rounde up to 16
+// (4 bytes unused)
+struct DrawDisplayListOp final : DLOp {
+  static const auto kType = DisplayListOpType::DrawDisplayList;
+
+  DrawDisplayListOp(const sk_sp<DisplayList> display_list)
+      : display_list(std::move(display_list)) {}
+
+  sk_sp<DisplayList> display_list;
+
+  void dispatch(Dispatcher& dispatcher) const {
+    dispatcher.drawDisplayList(display_list);
+  }
+};
+
+struct DrawTextBlobOp final : DLOp {
+  static const auto kType = DisplayListOpType::DrawTextBlob;
+
+  DrawTextBlobOp(const sk_sp<SkTextBlob> blob, SkScalar x, SkScalar y)
+      : x(x), y(y), blob(std::move(blob)) {}
+
+  const SkScalar x;
+  const SkScalar y;
+  const sk_sp<SkTextBlob> blob;
+
+  void dispatch(Dispatcher& dispatcher) const {
+    dispatcher.drawTextBlob(blob, x, y);
+  }
+};
+
+// struct DrawShadowRecOp final : DLOp {
```

Comment:
Commented out code.

Reply:
For a good reason documented in so many other places. I got tired of adding the description comment after a while.

Skia doesn't expose that structure (yet). There is a bug out there. They are looking into it. I cannot fully capture all SkCanvas rendering until I can support this structure. The code will eventually be used, and it must exist for completeness, but I am blocked by their header file structuring.

https://bugs.chromium.org/p/skia/issues/detail?id=12125

Reply:
(I'm hoping that they'll release the structure before I merge this. Otherwise I'll put more reasons in or switch these to ifdefs or something.)

Reply:
This Skia bug will be discussed in a synch-up today, so I can make a decision about this code soon.

Reply:
Skia says that this will take some time as they are also working on a redesign of the Shadow stuff so they would like to prioritize the Metal work over this header work. I think I'll just delete all references to the ShadowRec stuff (except, obviously, in the SkCanvas->DL adapter) for now. We only use this from Dart Canvas presently and that code already works around this issue.